### PR TITLE
feat(construction): Adds opaque and air boundary construction

### PR DESCRIPTION
**Pull request by tanushree04**
Addresses #175

### CODE REVIEW

File: lib/from_openstudio/material/opaque_no_mass.rb
Comment (tanushree04) [inline lib/from_openstudio/material/opaque_no_mass.rb:58]: For some reason, for the AirGap OS material the material.thermalAbsorptance, material.solarAbsorptance and material.visibleAbsorptance returns 0.0, even though the property isnt defined for it in OpenStudio. In this case the unless material.property.empty? statement was throwing an error and needed begin rescue statements.

File: lib/from_openstudio/material/opaque_no_mass.rb
Comment (tanushree04) [inline lib/from_openstudio/material/opaque_no_mass.rb:65]: This is adding defaults from the schema for MasslessOpaqueMaterial material if properties arent defined as well as AirGap

File: lib/from_openstudio/model.rb
Comment (tanushree04) [inline lib/from_openstudio/model.rb:65]: I realised that the construction is being created before other shade constructions from orphaned shade, room, face, aperture shades can be added to the $shade_construction hash and the method is called.

File: lib/from_openstudio/model.rb
Comment (tanushree04) [inline lib/from_openstudio/model.rb:74]: This should create shades constructions only after looping through the model

File: lib/from_openstudio/construction/shade.rb
Comment (tanushree04) [inline lib/from_openstudio/construction/shade.rb:56]: Im not sure why these properties are giving an error

File: lib/from_openstudio/material/opaque_no_mass.rb
Comment (chriswmackey) [reply lib/from_openstudio/material/opaque_no_mass.rb:58]: Interesting. That seems like a bug in OpenStudio SDK but I agree the right approach for now is just to work around it.  I think the current begin/rescue might be too implicit and generic that it could fail (or fail silently) after a fix is merged into OpenStudio SDK.

Couldn't we do just one check to see if the material isn't an AirGap and, if so, then we try to get the roughness, absorptances, etc.?

File: lib/from_openstudio/material/opaque_no_mass.rb
Comment (chriswmackey) [reply lib/from_openstudio/material/opaque_no_mass.rb:65]: We can just get rid of this part. [These fields are all optional](https://www.ladybug.tools/honeybee-schema/model.html#tag/energymaterial_model) and the core Python libraries automatically fill in the default when they aren't specified in the JSON. In fact, I think this might be a case where "less is more" and we should just keep the total amount of code we need to maintain to a minimum.

File: lib/from_openstudio/construction/air.rb
Comment (chriswmackey) [inline lib/from_openstudio/construction/air.rb:49]: I agree that a `TODO` is the right approach for now. This one needs to come from the EnergyPlus air mixing objects if we eventually implement it.

File: lib/from_openstudio/construction/shade.rb
Comment (chriswmackey) [inline lib/from_openstudio/construction/shade.rb:48]: I am unclear on why you are looping through the layers here. We only need to check the outer-most layer and all of the other layers that you are looping through have no effect on the resulting shade construction.  The other checks you have here about glazing vs. opaque materials looks good.  However, I think NoMass materials will also be a fairly common way that people specify the reflectance/absorptance of shades so we should also check for those.

File: lib/from_openstudio/geometry/room.rb
Comment (chriswmackey) [inline lib/from_openstudio/geometry/room.rb:104]: I feel like `$shade_construction` should be a hash where you organize the shade constructions by their unique name or ID and not a list with potentially a bunch of duplicated construction objects.  It's just going to make the run time a lot longer if you have all of these duplicated references to constructions in the list instead of one reference per unique construction. Also, is there a reason why you are doing this same copy/pasted check for each parent object (Room, Face, Aperture) instead of just having one check in the Shade class?

File: lib/from_openstudio/model.rb
Comment (chriswmackey) [reply lib/from_openstudio/model.rb:65]: Is this a question? Or a note that this code does not work?

In any event, I agree that `$shade_construction` should be a hash as you say and not a list as it is currently written in the code.

File: lib/from_openstudio/model.rb
Comment (chriswmackey) [inline lib/from_openstudio/model.rb:150]: Again, I feel like this is a sign that something hasn't been coded in the right place when I see it copy/pasted 5 times.

File: lib/from_openstudio/model.rb
Comment (chriswmackey) [inline lib/from_openstudio/model.rb:219]: Nice. That's a useful property of the OpenStudio SDK. However, I guess that you still have to check for the case that it's an opaque door construction (that still counts as fenestration). Maybe we should just forego this check and only stick to check the nature of the material layers.

File: lib/from_openstudio/model.rb
Comment (chriswmackey) [inline lib/from_openstudio/model.rb:221]: You don't need to do a whole loop here. Just grab the first/outermost material layer and check if it's Standard Glazing or Simple Glazing. It's illegal in E+ to have a gas gap on the outer layer of a window construction so you don't need to check for that material type if you are only checking the first layer.

File: lib/from_openstudio/model.rb
Comment (chriswmackey) [inline lib/from_openstudio/model.rb:231]: Again, you don't need to loop through all of these. Just grab the first layer. Also, I think you'll need to do a check for NoMass materials as well as Standard Opaque Materials.  You don't need to check for AirGap materials if you only check the first/outer-most layer since it's illegal to have them on the outside of a Construction in E+.

**Review by chriswmackey — CHANGES_REQUESTED** [review]
Thanks @tanushree04 ,

It's generally looking good but there are a few things that need to be changed before we merge. Also, we need to get the tests to pass. I can see that there's only one case that is failing now.  It looks like you are checking for the `frontSideSolarReflectanceatNormalIncidence` for a material that does not have this property:

https://github.com/ladybug-tools/honeybee-openstudio-gem/pull/206/checks?check_run_id=2723286240#step:3:342

File: lib/from_openstudio/geometry/room.rb
Comment (tanushree04) [reply lib/from_openstudio/geometry/room.rb]: Thanks Chris, that is a good suggestion.

File: lib/from_openstudio/construction/shade.rb
Comment (tanushree04) [reply lib/from_openstudio/construction/shade.rb:48]: In this case, I looped through all layers and am getting only the outermost, since we were looping through them anyway to get the namestring for each layer.

File: lib/from_openstudio/construction/shade.rb
Comment (tanushree04) [inline lib/from_openstudio/construction/shade.rb:64]: All of these commented out properties are throwing an Openstudio `undefined method` error. These methods do exist for the respective material types as boost optionals.


https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-3.1.0-doc/model/html/classopenstudio_1_1model_1_1_standard_glazing.html

https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-3.1.0-doc/model/html/classopenstudio_1_1model_1_1_standard_opaque_material.html

File: lib/from_openstudio/construction/shade.rb
Comment (tanushree04) [reply lib/from_openstudio/construction/shade.rb]: The NoMass materials do not have a reflectance property in OS.

https://openstudio-sdk-documentation.s3.amazonaws.com/cpp/OpenStudio-3.1.0-doc/model/html/classopenstudio_1_1model_1_1_massless_opaque_material.html

Comment (chriswmackey) [reply lib/from_openstudio/construction/shade.rb]: But `NoMass` material layers have a `solarAbsorptance`, which is always equal to `1 - solar reflectance`.  If you want to do a test to see if `NoMass` materials can't be used to Shades, then we can forgo the check. Otherwise, I'm inclined to think that they are permissible.

**Review by chriswmackey — APPROVED** [review]
This looks great, @tanushree04 . Thanks again for getting this together. I'll merge it and then let you know if anything comes up during our testing.

**Pull request merged**
